### PR TITLE
feat: add "postversion" scripts to automate publishing

### PR DESCRIPTION
**Pull request by wincent**
I have been doing a lot of [releases](https://github.com/liferay/liferay-npm-tools/releases) of late — about 20 in the last month — so I wanted to finally go ahead and add some more automation here which should make the release process both quicker and less error-prone.

This implements the last of the ideas in #50. (Previously we implemented a "preversion" script and ".yarnrc" files to create the appropriate release commits and tags.)

We add a "postversion" script that basically does this:

    # Check you are on the master branch
    git rev-parse --abbrev-ref HEAD

    # Check worktree is clean
    git diff --quite

    # Update upstream "master"
    git push upstream master

    # Merge "master" into "stable"
    git checkout stable
    git merge --ff-only master

    # Update upstream "stable"
    git push upstream stable --follow-tags

    # Return to the "master" branch
    git checkout master

    # Actually publish
    yarn publish

Obviously with abundant checks and guards and prompts along the way.

Test plan: I've sort-of-kind-of tested this as best I can without actually doing a release (ie. by commenting out stuff, or adding `--dry-run` to the `push` invocations, or by not supplying a OTP token to `yarn publish` etc). It might work. Maybe.

Closes: https://github.com/liferay/liferay-npm-tools/issues/50

automation

### CODE REVIEW

**Comment by wincent** [comment]
@jbalsas suggested we merge this and then we'll get somebody else (not me) to prepare the next release. We can watch it fail together. 😂

**Comment by wincent** [comment]
I had to release a new version of liferay-npm-scripts this morning but there was nobody in the office with publish access to test it. Will do a guinea pig test with somebody next time!

For what it's worth, it went swimmingly. Last PR to go in was this one: https://github.com/liferay/liferay-npm-tools/pull/177

This is what I saw when I ran `yarn version --patch` — first you see the "preversion output":

```
yarn version v1.16.0
info Current version: 4.12.0
$ cd ../.. && yarn ci
yarn run v1.16.0
$ yarn format:check && yarn lint && yarn test
$ prettier --list-different '**/*.js' '**/*.json' '**/*.md'
$ eslint '**/*.js'
$ jest
 PASS  packages/liferay-npm-scripts/test/utils/preprocessGlob.js
 PASS  packages/liferay-npm-scripts/test/index.js
 PASS  packages/liferay-npm-scripts/test/utils/getRegExpForGlob.js
 PASS  packages/liferay-npm-scripts/test/utils/deepMerge.js
 PASS  packages/liferay-npm-scripts/test/utils/getMergedConfig.js
 PASS  packages/liferay-npm-scripts/test/utils/spawnSync.js
 PASS  packages/liferay-npm-scripts/test/scripts/theme.js
 PASS  packages/liferay-npm-scripts/test/utils/expandGlobs.js
 PASS  packages/liferay-npm-scripts/test/scripts/webpack.js
 PASS  packages/liferay-npm-scripts/test/utils/spawnMultiple.js
 PASS  packages/liferay-npm-scripts/test/utils/findRoot.js
 PASS  packages/liferay-npm-scripts/test/utils/filterGlobs.js
 PASS  packages/liferay-jest-junit-reporter/test/index.js
 PASS  packages/liferay-npm-scripts/test/scripts/lint.js
 PASS  packages/liferay-npm-scripts/test/scripts/format.js
  ● Console

    console.log packages/liferay-npm-scripts/src/utils/log.js:9
      Prettier checked 1 file, found 0 files with problems, fixed 1 file


Test Suites: 15 passed, 15 total
Tests:       107 passed, 107 total
Snapshots:   2 passed, 2 total
Time:        2.703s
Ran all test suites.
✨  Done in 6.53s.
```

Then you see the version update:

```
info New version: 4.12.1
```

Then you see the "postversion" output:

```
$ node ../../publish.js
Push to upstream-rw/master? [y/n] y
Merge "master" into "stable"? [y/n] y
Push to upstream-rw/stable? [y/n] y
Run `yarn publish`? [y/n] y
Please enter an OTP token or press ENTER: 428076


Done! ✅

You can sanity-check that the package is correctly listed here:

https://www.npmjs.com/package/liferay-npm-scripts


✨  Done in 41.85s.
```

And finally, this is what the state of the world looks like afterwards:

```
* 822fb71 (HEAD -> master, tag: liferay-npm-scripts/v4.12.1, upstream-rw/stable, upstream-rw/master, stable) liferay-npm-scripts/v4.12.1 (78 seconds ago) <Greg Hurrell>
*   2844717 (upstream/master) Merge pull request #177 from liferay/wincent/jest-windows (4 minutes ago) <Greg Hurrell>
|\
| * f3e2d1f (wincent/jest-windows) refactor(scripts): remove unnecessary path manipulation (28 minutes ago) <Greg Hurrell>
| * 4d6a6ad feat(scripts): don't leave config files lying around on failure (28 minutes ago) <Greg Hurrell>
| * f40b8a2 fix: Revert "refactor(scripts): simplify Jest config set-up" (#176) (29 minutes ago) <Greg Hurrell>
|/
*   b5d2ffa Merge pull request #175 from liferay/wincent/auto-publish (20 hours ago) <Greg Hurrell>
```

The release is visible both [here](https://github.com/liferay/liferay-npm-tools/releases/tag/liferay-npm-scripts%2Fv4.12.1) and [here](https://www.npmjs.com/package/liferay-npm-scripts).

**Comment by jbalsas** [comment]
I find this line a bit confusing 

```
 PASS  packages/liferay-npm-scripts/test/scripts/format.js
  ● Console

    console.log packages/liferay-npm-scripts/src/utils/log.js:9
      Prettier checked 1 file, found 0 files with problems, fixed 1 file
```

😂

**Comment by wincent** [comment]
Are you referring to the fact that we aren't suppressing console output or that the contents of the console output is confusing?

About the *contents* of the output: I made it match the ESLint behavior (which is that if it "fixes" an issue, the issue doesn't appear in the "problems" count, so it can end up saying things like "found 0 issues, fixed 10 issues"). I guess what it really means is not "found X issues" but "X issues remain". I found it confusing, but I went with consistency with ESLint instead.

We can probably reword it to make it less weird/confusing.

Normally with Prettier, you're either running in "fix" mode and will see something like:

> checked 100 files, found 0 files with problems, fixed 1 file.

or you're running in "check" mode and will see something like:

> checked 100 files, found 10 files with problems.

ie. if you're in "fix" mode, the "files with problems" count is (almost) always 0.

The one edge case is when you are in "fix" mode but there is some file that *can't* be fixed (and the only reason I know for that is that there is a syntax error or a Prettier bug that causes it to throw an exception); in that case you can see something like:

> checked 100 files, found 1 file with problems, fixed 10 files.

While we can't easily change the way ESLint counts/reports its errors/fixes etc (because it is hard-coded in the "reporter"), we can trivially change Prettier. Maybe something like:

- "fix" mode, with no problems found; also "check" mode with no problems found:
  - "checked X files, 0 files have problems"
- "fix" mode, with some problems found and fixed:
  - "checked X files, fixed Y files with problems"
- "fix" mode, with some problems found and fixed, and some unfixable problems (eg. syntax errors):
  - "checked X files, fixed Y files with problems, Z files have problems"
- "check" mode, with some problems found (both formatting issues and unfixable problems like syntax errors):
  - "checked X files, Y files have problems"

**Comment by jbalsas** [comment]
Hey @wincent, 

> Are you referring to the fact that we aren't suppressing console output or that the contents of the console output is confusing?

First one.

I was just taking a quick look, so what stood out as weird was:

- We're running some checks to make sure we're in a good spot to publish
- I see a log that says some files had to be fixed, so a commit would need to be made
- We proceeded with publishing anyways

Looking closer, it seems obvious that both `$ prettier --list-different '**/*.js' '**/*.json' '**/*.md'`
`$ eslint '**/*.js'` and that the other thing is the expected `jest` output so all is good. It just tripped me momentarily :)

**Comment by wincent** [comment]
I think we can suppress that output in the jest tests in order to avoid this kind of confusion. The reason I didn't do it before was because I've seen in other projects of ours the opposite problem (which is that we suppress too aggressively, and then you wonder why `console.log` doesn't work when you're trying to debug a test). I'll add a narrowly scoped stub to make the test output clean.

**Comment by jbalsas** [comment]
It's okay, it's not a big deal. I've been bit by the "where's my `console.log` output" before :)

Also, I was just a know-it-all bystander, so you shouldn't take me seriously :)

**Comment by wincent** [comment]
Fix is too easy to ignore; it's just one line: https://github.com/liferay/liferay-npm-tools/pull/178
